### PR TITLE
[functorch] add switch to enable autograd.Function

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88785
* __->__ #88784

This is mostly a debug or "if you know what you're doing" switch for
now. It is not public API.

Test Plan:
- new tests

### CODE REVIEW

File: aten/src/ATen/functorch/DynamicLayer.cpp [hunk 1]
```diff
@@ -83,6 +83,18 @@ static DynmetaData& getGlobalDynmetaData() {
   return kDynMetaDataSingleton;
 }
 
+namespace {
+thread_local bool kIsAutogradFunctionAllowed = false;
```
Comment (samdow) [inline aten/src/ATen/functorch/DynamicLayer.cpp:87]: I would expect that if I enabled autograd.Function in one thread, all spawned threads would also have autograd.Function enabled (not certain if this will come up, but thinking that it might because of multithreading of the backward pass combined with higher order gradients). This version means that all threads will end up starting with the false behavior, right?

File: aten/src/ATen/functorch/DynamicLayer.h [hunk 2]
```diff
@@ -113,6 +113,12 @@ TORCH_API Tensor unwrapIfDead(const Tensor& tensor);
 TORCH_API std::ostream& operator<<(std::ostream& os, const DynamicLayer& layer);
 TORCH_API std::ostream& operator<<(std::ostream& os, const std::vector<DynamicLayer>& dynamicLayerStack);
 
+// While a functorch grad transform is active, autograd.Function is disabled
```
Comment (samdow) [inline aten/src/ATen/functorch/DynamicLayer.h:116]: nit: Isn't it all functorch transforms? Since vmap is a problem if the input tensor requires grad

File: aten/src/ATen/functorch/DynamicLayer.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/functorch/DynamicLayer.cpp:87]: You're right, ATen/ThreadLocalState needs to understand this. Thanks for catching

File: aten/src/ATen/functorch/DynamicLayer.h (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/functorch/DynamicLayer.h:116]: Yup, you're right. Will update

**Review by samdow — APPROVED** [review]
Awesome thanks for the updates!

File: test/functorch/test_eager_transforms.py [hunk 3]
```diff
@@ -2388,6 +2388,35 @@ def f(x):
...
+            def forward(ctx, x):
+                ctx.save_for_backward(x)
+                return x.sin()
+
+            @staticmethod
+            def backward(ctx, gy):
+                x, = ctx.saved_tensors
+                return gy * x.cos()
+
+        x = torch.randn([])
```
Comment (soulitzer) [inline test/functorch/test_eager_transforms.py:2403]: Is this tensor intentionally doesn't require gradients? I guess it doesn't matter because should error anyway.

**Review by soulitzer — APPROVED** [review]

File: test/functorch/test_eager_transforms.py (hunk 3 above)
Comment (zou3519) [reply test/functorch/test_eager_transforms.py:2403]: The functorch grad transform doesn't use the requires_gradness of the tensor to compute gradients, it accepts an argnums argument that says which input to compute gradients with respect to

Comment (soulitzer) [reply test/functorch/test_eager_transforms.py:2403]: Oh whoops, good point.
